### ISSUE TITLE
[Feat] Allow to configure returning parameter of entities

### ISSUE THREAD

**Issue by luchillo17** [typeorm/typeorm#978]
Currently if you make any operation it asks the DB to return only the id, no matter if creating or updating an entity.

Sometimes it's better to get the full entity values, like a source of truth, say you create a new item, you pass the values to TypeORM and only get the id, that's not that big of a deal since you know all data of that entity you send it, so you know the only thing you get back that you don't have is the id.

However let's say you update an entity with partial data, say the User entity you updated it's password, then you want the DB to return the full entity but the password, so that you can save that directly into your client db as the latest source of truth about the state of the entity on server.

If i were to do the above, currently i would have to `await` for the update to happen and then query the database for the item with the id returned by the update operation, that's just not scalable because every operation would be composed of minimum 2 calls to DB.

Instead of just returning `id` on create and an empty resolved promise on update, i think we should be able to ask TypeORM to return certain attributes in both operations.

**Comment by pleerock** [comment]
ahhh dude its sooo hard for me to understand your requests, always... sorry about that, but can you say simply what is the problem (on example). I understood that you want to be able to return more stuff after insertion, but I would like to understand what why. I didn't understand example with password.

**Comment by luchillo17** [comment]
Oh sorry, let me try again:
**Case 1:** Say you insert an user, if operation is successful you save in client, if you save previous to the operation you need to delete from local, the idea is to save the actual result from the server, like using the DB as the single source of truth, so i would like to return the full entity or part of it, not just the **Id**.

**Case 2:** There's some data that get's generated on server, like default values, created and updated dates, entity version... , without allowing **returning** configuration, we are forced to query the recently persisted Entity to get those values and return back to the client.

**Case 3:** Sequelize allows it, TypeORM should as well.

Is that better or worse?

**Comment by iz-iznogood** [comment]
The query builder allows for a returning clause where it can be applied (postgres, mssql).
Why don't you use this ? I tried and it works great

**Comment by luchillo17** [comment]
Whhaaattt? Sure didn't saw that on the docs.

**Comment by luchillo17** [comment]
Ok so the QueryBuilder does, but the Entity manager doesn't, and it's very verbose if i have to write to retur each column, in my case i would like to return all fields by default.

**Comment by luchillo17** [comment]
Also typings doesn't show me the type definitions of the clause, how would that be called?

**Comment by iz-iznogood** [comment]
How about this ?

```ts
        const savedEntities = await connection.getRepository("post")
            .createQueryBuilder("post")
            .update({
                col1: "123"
            })
            .returning("inserted.*");
```

Is this not enough ?

**Comment by luchillo17** [comment]
I'm trying it but seems i lose auto-generated columns like `version` or `createdAt` from the entity class.

```ts
const controlPlace = new ControlPlace()
    Object.assign(controlPlace, data)
    controlPlace.createdBy = controlPlace.updatedBy = Promise.resolve(user);
    return connection
      .createQueryBuilder(ControlPlace, 'controlplace')
      .insert()
      .values(controlPlace)
      .returning('*')
      .execute()
```

If it breaks for those columns in creation, i think maybe the `update` doesn't update those fields when it's called.

**Comment by iz-iznogood** [comment]
Can you create a simple test so that I can test this ?
In my case it worked great. 
What db are you using ?

**Comment by luchillo17** [comment]
Postgres, most likely you don't have auto-generate columns? Here's how the entity looks like:
```ts
@Entity('MON_CONTROL_PLACE', {
  orderBy: {
    name: 'ASC',
    id: 'DESC',
  },
} as any)
export class ControlPlace {
  @PrimaryGeneratedColumn({
    // name: 'MON_CONTROL_PLACE_ID',
  })
  id: number

  @Column({
    nullable: false,
    unique: true,
  })
  name: string

  // Relations
  @OneToMany(
    (type) => RouteLine,
    (routeLine) => routeLine.controlPlace,
  )
  routeLines: Promise<RouteLine[]>

  @ManyToOne(
    (type) => User,
    (user) => user.createdControls,
    { nullable: false },
  )
  @JoinColumn()
  createdBy: Promise<User>

  @ManyToOne(
    (type) => User,
    (user) => user.updatedControls,
    { nullable: false },
  )
  @JoinColumn()
  updatedBy: Promise<User>

  /**
   * Auto-generated at insert
   *
   * @type {Date}
   * @memberof ControlPlace
   */
  @CreateDateColumn()
  createdAt: Date

  /**
   * Auto-updated at insert & update
   *
   * @type {Date}
   * @memberof ControlPlace
   */
  @UpdateDateColumn()
  updatedAt: Date

  /**
   * Version of entity item, updated on update
   *
   * @type {number}
   * @memberof ControlPlace
   */
  @VersionColumn()
  version: number
}

```

**Comment by iz-iznogood** [comment]
Indeed I tried it and you are right.
Also query builder does not update the autogenerated columns, something is missing there

**Comment by pleerock** [comment]
Of course @luchillo17 can use InsertQueryBuilder, but it will never replace him `save` method, because InsertQueryBuilder is for trivial operations, and its intended to prevent user from extra computations. If we continue to add things to it will may become same complex as `save` method. This time we need update functionality, next time we'll need something else `save` method contains.

For now I suggest you to re-load your entity. 
Of course we can re-load entity once we saved it, but its another extra request, so Im not really sure if this requirement worth it. But I understand that you would like to have values generated by `default`, if we reload entity it could resolve issue with update and create dates as well. Let's left it as discussion and maybe in the future we'll implement re-loading entities after save. The most important is we should determine if new request really worth it.

**Comment by iz-iznogood** [comment]
@pleerock QueryBuilder IMHO needs to implement at least the minimum functionality by updating the audit columns and the concurrency checking logic (version column update).

All other functionality is not necessary since updating graphs can be implemented outside the query builder as a class or library. 
But the insert - update code must behave the same for a single entity. From what I have seen the complexity is on the code that updates the relations - hierarchies. 

Also the cost on performing a fetch to get the saved entity varies on the scenario, in some cases it can be substantial.

**Comment by pleerock** [comment]
there are lot of other operations going on in `save` method as well. Example are subscribers and listeners. We will need to include them as well. I don't remember all such small operations right now happening in save method.

**Comment by iz-iznogood** [comment]
I will take a look and will come back to this if you agree. 
I believe that the code that guarantees data integrity should be included. All else is optional, at least for now.

**Comment by pleerock** [comment]
I have re-worked how persistence works, so this request is now completely covered:

> There's some data that get's generated on server, like default values, created and updated dates, entity version... , without allowing returning configuration, we are forced to query the recently persisted Entity to get those values and return back to the client.

This change is available in `next` version of typeorm, you can try `npm i typeorm@next`
